src/apps/btc/btc_common.c

```diff
@@ -403,3 +439,46 @@ bool btc_common_pkscript_from_multisig(
 
     return true;
 }
+
+USE_RESULT bool btc_common_multisig_is_valid(
+    const BTCScriptConfig_Multisig* multisig,
+    const uint32_t* keypath,
+    size_t keypath_len,
+    uint32_t expected_coin,
+    size_t* our_xpub_index_out)
+{
+    if (multisig->xpubs_count > MULTISIG_P2WSH_MAX_SIGNERS) {
+        return false;
+    }
+    if (multisig->threshold == 0 || multisig->threshold > multisig->xpubs_count) {
+        return false;
+    }
+    if (!_is_valid_keypath_account_multisig_p2wsh(keypath, keypath_len, expected_coin)) {
+        return false;
+    }
+
+    struct ext_key our_xpub __attribute__((__cleanup__(keystore_zero_xkey))) = {0};
+    if (!keystore_get_xpub(keypath, keypath_len, &our_xpub)) {
+        return false;
+    }
+
+    bool our_xpub_found = false;
+    for (size_t cosigner_idx = 0; cosigner_idx < multisig->xpubs_count; cosigner_idx++) {
```

Comment:
This seems a bit inefficient. How about adding the index of our key to the multisig message and only do the iteration if the memcmp between the key at that index and ours fails?

Reply:
by key you mean xpub right? key alone is not enough. memcmp doesn't work because of different types (hence `apps_common_bip32_xpubs_equal`).

Index provided by client would make it one comparison instead of looping and comparing all of them, true. Efficiency wise I wouldn't worry, it's a simple operation and isn't noticeable. The client integrations would be slightly more complicated, but on the other hand the rule that at least one has to be ours would be encoded in the message, which is nice.

I think I'll do it :P thanks!

Reply:
Done now.

Reply:
Actually this simplified this function and the commits that build on top quite a bit, nice suggestion!